File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let left = 0;
    let count = 0;
    let hashmap = new Map();

    for(let right = 0 ; right< s.length; right++){
        if(hashmap.has(s[right]))
        hashmap.set(s[right], hashmap.get(s[right])+1);

        else
        hashmap.set(s[right], 1);

        while(hashmap.get(s[right]) > 1){
            hashmap.set(s[left], hashmap.get(s[left])-1);
            left++;
        }

        count = Math.max(count, right - left + 1);

    }

    return count;
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int count = 0;
        int result = 0;
        int low = 0;
        int high = 0;

        while(high<s.length()){
            while(map.containsKey(s.charAt(high))){
                map.remove(s.charAt(low));
                low++;
            }


            map.put(s.charAt(high), 1);

            result = Math.max(result, high-low+1);
            high++;
        }

        return result;
    }
}